Framework/CurveFitting/src/Functions/Meier.cpp

```diff
@@ -0,0 +1,184 @@
+                   "exerted by the presence of the muon");
+  declareParameter("Sigma", 0.2, "Gaussian decay rate");
+  declareParameter("Lambda", 0.1, "Exponential decay rate");
+  // J, Total angular momentum quanutm number
+  declareAttribute("Spin", API::IFunction::Attribute(3.5));
+}
+
+void Meier::function1D(double *out, const double *xValues, const size_t nData) const {
+  const double A0 = getParameter("A0");
+  const double J2 = round(getAttribute("Spin").asDouble() * 2);
+  const double Lambda = getParameter("Lambda");
+  const double Sigma = getParameter("Sigma");
+
+  std::valarray<double> xValArray(xValues, nData);
+
+  const std::valarray<double> gau = std::exp(-0.5 * pow((Sigma * xValArray), 2));
+  const std::valarray<double> Lor = std::exp(-Lambda * xValArray);
+
+  std::valarray<double> positiveLambda;
+  std::valarray<double> negativeLambda;
+  std::valarray<double> cos2AlphaSquared;
+  std::valarray<double> sin2AlphaSquared;
+  std::valarray<double> cosAlphaSquared;
+  std::valarray<double> sinAlphaSquared;
+
+  precomputeIntermediateSteps(sinAlphaSquared, cosAlphaSquared, sin2AlphaSquared, cos2AlphaSquared, negativeLambda,
+                              positiveLambda, J2);
+
+  std::valarray<double> Px;
+  calculatePx(Px, xValArray, sinAlphaSquared, cosAlphaSquared, negativeLambda, positiveLambda, J2);
+
+  std::valarray<double> Pz;
+  calculatePz(Pz, xValArray, sin2AlphaSquared, cos2AlphaSquared, negativeLambda, positiveLambda, J2);
+
+  const std::valarray<double> outValArray = A0 * gau * Lor * (1. / 3.) * (2 * Px + Pz);
+  std::copy(begin(outValArray), end(outValArray), out);
+}
+
+/**
+ * Precomputes intermediate terms used to calculate the polrization in the x and z directions. All value arrays will be
+ * resized to J2 + 2 and set to their respective quantities
+ * @param sinAlphaSquared :: sin of alpha squared
+ * @param cosAlphaSquared :: cos of alpha squared
+ * @param sin2AlphaSquared :: sin of 2*alpha squared
+ * @param cos2AlphaSquared :: cos of 2*alpha squared
+ * @param positiveLambda :: negative lambda
+ * @param negativeLambda :: positive lambda
+ * @param J2 :: 2 * total angular momentum quantum number
+ */
+void Meier::precomputeIntermediateSteps(std::valarray<double> &sinAlphaSquared, std::valarray<double> &cosAlphaSquared,
+                                        std::valarray<double> &sin2AlphaSquared,
+                                        std::valarray<double> &cos2AlphaSquared, std::valarray<double> &negativeLambda,
+                                        std::valarray<double> &positiveLambda, const double &J2) const {
+  const double FreqD = getParameter("FreqD");
+  const double FreqQ = getParameter("FreqQ");
+  const double J = J2 / 2;
+
+  const double OmegaD = 2 * M_PI * FreqD;
+  const double OmegaQ = 2 * M_PI * FreqQ;
+
+  const size_t size = int(J2 + 2);
+  sinAlphaSquared.resize(size);
```

Comment:
why are these being passed into the function then assigned to, instead of being returned?